File: Solutions/1882. Process Tasks Using Servers/1882. Process Tasks Using Servers.cpp
//fast,but unclear version
typedef array<int,3> ar;
class Solution {
public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        //min heap
        priority_queue<ar,vector<ar>,greater<ar>> avail,busy;
        vector<int> res;
        // time,weight,index
        for(int i=0;i<servers.size();++i)
            avail.push({0,servers[i],i});
        for(int t=0;t<tasks.size();++t){
            while(!busy.empty() and (busy.top()[0]<=t or avail.empty())){
                auto [time,w,i]=busy.top();busy.pop();
                avail.push({time<=t?0:time,w,i});
            }
            auto [time,w,s]=avail.top();avail.pop();
            busy.push({max(time,t)+tasks[t],w,s});
            res.push_back(s);
        }
        return res;
    }
};

//clear version
typedef array<int,3> ar;
class Solution {
public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        vector<int> res;
        priority_queue<ar,vector<ar>,greater<ar>> avail,busy;
        for(int i=0;i<servers.size();++i)
            avail.push({servers[i],i,0});
        for(int i=0;i<tasks.size();++i){
            int t=tasks[i];
            //try to free servers
            while(busy.size()>0 and busy.top()[0]<=i){
                auto [time,weight,index]=busy.top();busy.pop();
                avail.push({weight,index,time});
            }
            if(avail.size()==0){
                auto [time,weight,index]=busy.top();busy.pop();
                res.push_back(index);
                time+=t;
                busy.push({time,weight,index});
            }else{
                auto [weight,index,time]=avail.top();avail.pop();
                res.push_back(index);
                time=i+t;
                busy.push({time,weight,index});
            }
        }
        return res;
    }
};

File: Solutions/1882. Process Tasks Using Servers/1882. Process Tasks Using Servers.py
class Solution:
    def assignTasks(self, servers: List[int], tasks: List[int]) -> List[int]:
        res=[]
        avail=[[weight,i,0] for i,weight in enumerate(servers)]
        heapq.heapify(avail)
        busy=[]
        for i in range(len(tasks)):
            t=tasks[i]
            #try to free server
            while busy and busy[0][0]<=i:
                time,weight,index=heapq.heappop(busy)
                heapq.heappush(avail,[weight,index,time])
            if len(avail)==0:
                time,weight,index=heapq.heappop(busy)
                res.append(index)
                time+=t
                heapq.heappush(busy,[time,weight,index])
            else:
                weight,index,time=heapq.heappop(avail)
                res.append(index)
                time=t+i
                heapq.heappush(busy,[time,weight,index])
        return res
            